File: android/app/src/main/java/com/mewo/reader/ui/reader/TimelinePost.kt
package com.mewo.reader.ui.reader

import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.ChatBubbleOutline
import androidx.compose.material.icons.outlined.Favorite
import androidx.compose.material.icons.outlined.FavoriteBorder
import androidx.compose.material.icons.outlined.IosShare
import androidx.compose.material.icons.outlined.Repeat
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.scale
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.role
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import com.mewo.reader.data.FeedPost
import com.mewo.reader.data.PostKind
import com.mewo.reader.ui.components.Avatar
import com.mewo.reader.ui.theme.LocalMewoColors
import com.mewo.reader.ui.theme.LocalReaderFont
import java.io.File

@Composable
fun TimelinePost(
    post: FeedPost,
    author: String,
    handle: String,
    cover: File?,
    liked: Boolean,
    commentCount: Int = 0,
    commented: Boolean = false,
    onComment: () -> Unit,
    onLike: () -> Unit,
    onRepost: () -> Unit,
    onShare: () -> Unit,
    onOpen: (() -> Unit)? = null,
    onChapterClick: (() -> Unit)? = null,
) {
    val isHeading = post.kind == PostKind.Heading
    val readerType = LocalReaderFont.current.scale
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .then(
                if (onOpen != null) Modifier.clickable(role = Role.Button, onClick = onOpen)
                else Modifier,
            )
            .padding(start = 16.dp, end = 12.dp, top = 12.dp),
    ) {
        Avatar(name = author, coverFile = cover, size = 40.dp)
        Spacer(Modifier.width(12.dp))
        Column(Modifier.weight(1f)) {
            Row(verticalAlignment = Alignment.CenterVertically) {
                Text(
                    text = author,
                    style = MaterialTheme.typography.titleSmall,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                    modifier = Modifier.weight(1f, fill = false),
                )
                Spacer(Modifier.width(6.dp))
                Text(
                    text = "@$handle · ${post.chapter}",
                    style = MaterialTheme.typography.labelSmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                    modifier = Modifier
                        .weight(1f)
                        .then(
                            if (onChapterClick != null) {
                                Modifier
                                    .clickable(role = Role.Button, onClick = onChapterClick)
                                    .semantics {
                                        contentDescription = "Chapters, ${post.chapter}"
                                        role = Role.Button
                                    }
                            } else {
                                Modifier
                            },
                        ),
                )
            }
            Spacer(Modifier.height(4.dp))
            Text(
                text = post.text,
                style = if (isHeading) {
                    readerType.headingStyle(MaterialTheme.typography.titleMedium)
                } else {
                    readerType.bodyStyle(MaterialTheme.typography.bodyLarge)
                },
            )
            Spacer(Modifier.height(4.dp))
            PostActions(
                liked = liked,
                commentCount = commentCount,
                commented = commented,
                onComment = onComment,
                onLike = onLike,
                onRepost = onRepost,
                onShare = onShare,
            )
            Spacer(Modifier.height(4.dp))
        }
    }
    HorizontalDivider(color = MaterialTheme.colorScheme.outline, thickness = 0.6.dp)
}

@Composable
fun PostActions(
    liked: Boolean,
    commentCount: Int,
    commented: Boolean,
    onComment: () -> Unit,
    onLike: () -> Unit,
    onRepost: () -> Unit,
    onShare: () -> Unit,
) {
    val mute = MaterialTheme.colorScheme.onSurfaceVariant
    val likeColor = LocalMewoColors.current.like
    val accent = MaterialTheme.colorScheme.primary
    val likeScale by animateFloatAsState(if (liked) 1.18f else 1f, label = "like")
    Row(
        modifier = Modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.SpaceBetween,
        verticalAlignment = Alignment.CenterVertically,
    ) {
        ActionIcon(
            icon = Icons.Outlined.ChatBubbleOutline,
            tint = if (commented) accent else mute,
            label = if (commentCount == 0) {
                "Comment on this line"
            } else {
                "Comments, $commentCount"
            },
            count = commentCount,
            onClick = onComment,
        )
        ActionIcon(
            icon = Icons.Outlined.Repeat,
            tint = mute,
            label = "Quote this line",
            onClick = onRepost,
        )
        ActionIcon(
            icon = if (liked) Icons.Outlined.Favorite else Icons.Outlined.FavoriteBorder,
            tint = if (liked) likeColor else mute,
            label = if (liked) "Unlike" else "Like",
            onClick = onLike,
            modifier = Modifier.scale(likeScale),
        )
        ActionIcon(
            icon = Icons.Outlined.IosShare,
            tint = mute,
            label = "Share",
            onClick = onShare,
        )
    }
}

@Composable
private fun ActionIcon(
    icon: ImageVector,
    tint: Color,
    label: String,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    count: Int = 0,
) {
    IconButton(
        onClick = onClick,
        modifier = modifier
            .size(width = if (count > 0) 52.dp else 40.dp, height = 40.dp)
            .semantics {
                contentDescription = label
                role = Role.Button
            },
    ) {
        Row(verticalAlignment = Alignment.CenterVertically) {
            Icon(icon, contentDescription = null, tint = tint, modifier = Modifier.size(18.dp))
            if (count > 0) {
                Spacer(Modifier.width(4.dp))
                Text(
                    text = count.toString(),
                    style = MaterialTheme.typography.labelSmall,
                    color = tint,
                )
            }
        }
    }
}

File: ios/Mewo/UI/Reader/TimelinePost.swift
import SwiftUI

struct TimelinePost: View {
    @Environment(\.mewo) private var mewo
    @Environment(\.readerScale) private var scale
    var post: FeedPost
    var author: String
    var handle: String
    var cover: URL?
    var liked: Bool
    var commentCount: Int = 0
    var commented: Bool = false
    var onComment: () -> Void
    var onLike: () -> Void
    var onRepost: () -> Void
    var onShare: () -> Void
    var onOpen: (() -> Void)?
    var onChapterClick: (() -> Void)?

    var body: some View {
        VStack(spacing: 0) {
            HStack(alignment: .top, spacing: 12) {
                Avatar(name: author, cover: cover, size: 40)
                VStack(alignment: .leading, spacing: 4) {
                    HStack(spacing: 6) {
                        Text(author)
                            .font(MewoFont.bold(15))
                            .lineLimit(1)
                        chapterLine
                    }
                    Text(post.text)
                        .font(post.kind == .heading ? MewoFont.bold(scale.headingSize) : MewoFont.regular(scale.bodySize))
                        .lineSpacing(max(0, (post.kind == .heading ? scale.headingLine : scale.bodyLine) - (post.kind == .heading ? scale.headingSize : scale.bodySize)))
                        .foregroundStyle(mewo.ink)
                    PostActions(
                        liked: liked,
                        commentCount: commentCount,
                        commented: commented,
                        onComment: onComment,
                        onLike: onLike,
                        onRepost: onRepost,
                        onShare: onShare
                    )
                }
            }
            .padding(.leading, 16)
            .padding(.trailing, 12)
            .padding(.top, 12)
            .contentShape(Rectangle())
            .onTapGesture { onOpen?() }
            Hairline()
        }
        .foregroundStyle(mewo.ink)
        .accessibilityIdentifier("post-\(post.id)")
        .accessibilityElement(children: .contain)
    }

    private var chapterLine: some View {
        Text("@\(handle) · \(post.chapter)")
            .font(MewoFont.regular(13))
            .foregroundStyle(mewo.mute)
            .lineLimit(1)
            .onTapGesture { onChapterClick?() }
            .accessibilityLabel("Chapters, \(post.chapter)")
    }
}

struct PostActions: View {
    @Environment(\.mewo) private var mewo
    var liked: Bool
    var commentCount: Int
    var commented: Bool
    var onComment: () -> Void
    var onLike: () -> Void
    var onRepost: () -> Void
    var onShare: () -> Void

    var body: some View {
        HStack {
            action(
                icon: "bubble.right",
                tint: commented ? mewo.accent : mewo.mute,
                label: commentCount == 0 ? "Comment on this line" : "Comments, \(commentCount)",
                count: commentCount,
                action: onComment
            )
            action(icon: "arrow.2.squarepath", tint: mewo.mute, label: "Quote this line", action: onRepost)
            action(
                icon: liked ? "heart.fill" : "heart",
                tint: liked ? mewo.like : mewo.mute,
                label: liked ? "Unlike" : "Like",
                scale: liked ? 1.18 : 1,
                action: onLike
            )
            action(icon: "square.and.arrow.up", tint: mewo.mute, label: "Share", action: onShare)
        }
    }

    private func action(icon: String, tint: Color, label: String, count: Int = 0, scale: CGFloat = 1, action: @escaping () -> Void) -> some View {
        Button(action: action) {
            HStack(spacing: 4) {
                Image(systemName: icon)
                    .font(.system(size: 16))
                    .scaleEffect(scale)
                if count > 0 {
                    Text("\(count)").font(MewoFont.regular(13))
                }
            }
            .foregroundStyle(tint)
            .frame(minWidth: count > 0 ? 52 : 40, minHeight: 40)
        }
        .buttonStyle(.plain)
        .accessibilityLabel(label)
    }
}

struct ShareSheet: UIViewControllerRepresentable {
    var text: String

    func makeUIViewController(context: Context) -> UIActivityViewController {
        UIActivityViewController(activityItems: [text], applicationActivities: nil)
    }

    func updateUIViewController(_ uiViewController: UIActivityViewController, context: Context) {}
}
